pypesto/sampling/pymc3.py

```diff
@@ -112,11 +105,221 @@ def get_samples(self) -> McmcPtResult:
+        else:
+            return 'free_parameters'
+    else:
+        return 'theta'
+
+
+def pymc3_to_arviz(model: pm.Model, trace: pm.backends.base.MultiTrace, problem: Problem = None, vectorize: bool = False):
+    kwargs = {}
+    if vectorize:
+        if problem is None:
+            raise ValueError('if vectorize is True, ' \
+                             'a problem must be given')
+        x_free_names = [problem.x_names[i] for i in problem.x_free_indices]
+        kwargs['coords'] = {"free_parameter": x_free_names}
+        kwargs['dims'] = {pymc3_vector_parname(problem): ["free_parameter"]}
+
+    return az.from_pymc3(
+        trace=trace,
+        model=model,
+        log_likelihood=False,  # NB at the moment we do not separate prior
+                               # from posterior, thus computing likelihood
+                               # from pymc3 model is meaningless
+        **kwargs
+    )
+
+
+def BetterUniform(name, *, lower, upper, lerp='convex', **kwargs):
+    """
+    A uniform bounded random variable whose behaviour near the boundary of
+    the domain is better than the native `pymc3.Uniform`.
+
+    The problem with `pymc3.Uniform` is that the inverse transform formula
+    fails in floating point arithmetic when the lower and the upper bound differ
+    largely in magnitude, resulting in a value outside the original
+    interval. This leads the log-posterior for the original distribution to
+    become `-inf`. By using a better formula this can be avoided.
+
+    While this is a sufficient fix, there is another opportunity for
+    simplification: avoiding the computation of `log(ub - lb)` in the
+    log-posterior (it appears both in the original log-posterior and in the
+    transform jacobian, so it can be simplified away).
+    This simplification is imposed by using instead of `Uniform` a `Flat` prior
+    (which has log-posterior 0) and removing the term `log(ub - lb)` from
+    the interval transformation jacobian.
+    """
+    if 'shape' not in kwargs.keys():
+        if 'testval' in kwargs.keys():
+            shape = np.broadcast(lower, upper, kwargs['testval']).shape
+        else:
+            shape = np.broadcast(lower, upper).shape
+        kwargs['shape'] = shape
+    scalar = (len(kwargs['shape']) == 0)
+
+    if 'transform' in kwargs.keys():
+        raise Exception('if specifying a custom transform, ' \
+                        'please use pymc3.Uniform')
+
+    BoundedFlat = pm.Bound(pm.Flat, lower=lower, upper=upper)
+
+    transform = BetterInterval(lower, upper, lerp, scalar=scalar)
+
+    return BoundedFlat(name, transform=transform, **kwargs)
+
+    # In the case we start from pm.Uniform,
+    # we need to comment the jacobian out of BetterInterval
+    # transform = BetterInterval(lower, upper, lerp, scalar=scalar)
+    # return pm.Uniform(name, lower=lower, upper=upper, transform=transform, **kwargs)
+
+
+class BetterInterval(pm.distributions.transforms.Interval):
```

Comment:
please move all this pmc3/theano/... specific code to a different file, like it was done for the TheanoLogPosterior etc.. Otherwise we will again run into problems when using pypesto without a pymc3 installation.

Reply:
So that it does not get imported automatically by the `__init__.py` file, right? I see I did not get that, I will move the low level interface to a different file.

Reply:
exactly. in the file which gets imported on module level (pymc3.py), every critical import should be try-excepted. and since you explicitly use hierarchical class names in function and class constructors, these parts should either by written as strings (not always possible) or hidden in a separate file, and then in pymc3.py you do a single try-excepted import of those things, like done for the TheanoLogProbability.

Reply:
But why are imports excepted also inside `theano.py`? If someone is trying to import `theano.py` and `theano` is not available, wouldn't an `ImportError` be the reasonable outcome?

Reply:
And would it be possible to try importing `theano` inside `__init__.py` before importing `pymc3.py`? So that the user cannot even try to instantiate the pymc3 sampler.

Reply:
> But why are imports excepted also inside `theano.py`? If someone is trying to import `theano.py` and `theano` is not available, wouldn't an `ImportError` be the reasonable outcome?

You can try it out. It may have lead to problems when building the docs or with pycharm, but worth a shot. One should never directly try to import theano.py, only via pymc3.py, as stand-alone the theano things have no meaning.

Reply:
> And would it be possible to try importing `theano` inside `__init__.py` before importing `pymc3.py`? So that the user cannot even try to instantiate the pymc3 sampler.

I think this will not be necessary, I'd rather keep the `__init__` clean. rather, when the `Pymc3Sampler.__init__` is called, availability of the packages can be check dynamically and an error raised otherwise.